s3_storage_provider.py

```diff
@@ -65,25 +70,32 @@ def __init__(self, hs, config):
         if "secret_access_key" in config:
             self.api_kwargs["aws_secret_access_key"] = config["secret_access_key"]
 
+        self._download_pool = ThreadPool(name="s3-download-pool", maxthreads=40)
+        self._download_pool.start()
+
     def store_file(self, path, file_info):
         """See StorageProvider.store_file"""
 
         def _store_file():
             session = boto3.session.Session()
-            session.resource('s3', **self.api_kwargs).Bucket(self.bucket).upload_file(
+            session.resource("s3", **self.api_kwargs).Bucket(self.bucket).upload_file(
                 Filename=os.path.join(self.cache_directory, path),
                 Key=path,
                 ExtraArgs={"StorageClass": self.storage_class},
             )
 
-        return make_deferred_yieldable(
-            reactor.callInThread(_store_file)
-        )
+        # XXX: reactor.callInThread doesn't return anything, so I don't think this does
+        # what the author intended.
+        return make_deferred_yieldable(reactor.callInThread(_store_file))
```

Comment:
Hmm, this function is meant to return a Deferred when we finished uploading the file. I guess this means the `store_synchronous` flag doesn't work

Reply:
Also, any reason not to use the thread pool we created?

Reply:
Aha, we probable want to use https://twistedmatrix.com/documents/current/api/twisted.internet.threads.html#deferToThreadPool

Reply:
yeah. All of this is somewhat separate to the crisis I was trying to deal with at 11pm the other night...